sdk/cosmos/azcosmos/async_cache.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"context"
+	"reflect"
+	"sync"
+)
+
+type invalidCacheValue struct{}
+
+func (i invalidCacheValue) Error() string { return "Invalid cache value" }
+
+type asyncCache struct {
+	values sync.Map
+}
+
+type cacheValueTask func() interface{}
+
+type cacheValue struct {
+	value    interface{}
+	complete bool
+	fn       cacheValueTask
+	ch       <-chan interface{}
+}
+
+func newAsyncCache() *asyncCache {
+	return &asyncCache{}
+}
+
+func (ac *asyncCache) set(key interface{}, value interface{}) {
+	ac.values.Store(key, cacheValue{value: value})
```

Comment:
Does the sync.Map take care of concurrent sets for the same key?

Reply:
sync.Map should take care of concurrent sets as it uses a RWMutex under the hood.